Use pact jest for more readable and cleaner tests

MatthiasKainer opened the pull request:
Adding pact-jest would allow cleaner tests in the js realms. Also, I split up responsibilities for the following ideas:

1. Split up request and response. In general, this makes the tests much more readable and reusable
2. Allow the provider to specify some defaults for common shared usecases of their consumers. Not every consumer has to necessarily reinvent the wheel (create their own product creators) here. Providing builders and type definition are other ideas on how a provider can share "convenience".  
3. Testing against a snapshot rather than a manual description. Pact will do the heavy lifting for us, we can focus on data. If we have certain specifics those can be added, but are not necessary on every field imho.

MatthiasKainer committed:
Use pact jest for more readable and cleaner tests

atothek1 commented on consumer-a/tests/getProducts.spec.ts:
```diff
@@ -1,61 +1,26 @@
-describe("consumer-a/getProducts()", () => {
+import { pactWith } from 'jest-pact';
+import { API, api } from "../src";
+import { productsRequest, withListOfProductsRequest } from "provider/src/endpoints";
```
feedback:

I intentially avoided cross package sharing as i wanted to treat each package as isolated and independant. Like in a distrubuted dev team, the provider might be written in a different language, the different consumer are distributed accross different repos etc.

MatthiasKainer replied:
jop, fully agree. As I tried to say in the description, this is a convenience thing that one might provide to some, not a requirement that has to be used by all

atothek1 replied:
> jop, fully agree. As I tried to say in the description, this is a convenience thing that one might provide to some, not a requirement that has to be used by all

okay, you mean even if the provider would be written in Java that they could nicely provide a JS package to be used for testing.

atothek1 commented on consumer-a/src/index.ts:
```diff
@@ -1,8 +1,18 @@
+    getProducts: () => Promise<Product[]>
+}
+
+export function api(baseUrl: string): API {
```
feedback:

In general i am not a fan of adjusting the source code to satisfy testing requirement, I think this is not the way to go. At least in my opinion.

MatthiasKainer replied:
You don't have to do this - I was thinking about adding a second endpoint `/products/:byId` to show a possibly more interesting usecase (with 201, 404, and https://http.cat/405) but I haven't had the time to start putting it in code. That was just the first refactoring for that step to be able to share the baseUrl and get that argument out of the way. I should propably have stashed it for a different commit, sorry

atothek1 replied:
> You don't have to do this - I was thinking about adding a second endpoint `/products/:byId` to show a possibly more interesting usecase (with 201, 404, and https://http.cat/405) but I haven't had the time to start putting it in code. That was just the first refactoring for that step to be able to share the baseUrl and get that argument out of the way. I should propably have stashed it for a different commit, sorry

yeah :) i have prepared it in cosnumer-c, but also not finished it ;)

atothek1 commented on consumer-a/tests/getProducts.spec.ts:
```diff
@@ -1,61 +1,26 @@
+import { API, api } from "../src";
+import { productsRequest, withListOfProductsRequest } from "provider/src/endpoints";
 
+pactWith({ consumer: 'e2e Consumer A', provider: 'e2e Provider' }, provider => {
```
feedback:

this looks nice, the abstraction with jest-pact makes sense, will adapt. Thanks

atothek1 commented on provider/src/endpoints/index.ts:
```diff
@@ -0,0 +1,23 @@
+import { Interaction, InteractionObject, Matchers, RequestOptions, ResponseOptions } from '@pact-foundation/pact';
```
feedback:

As I wrote above this only works within the same tech stack and a monorepo, not if the provider is written in python or java and in an own repository.

MatthiasKainer committed:
Added more scenarios in consumer-c

Also refactored provider to avoid setup/teardown logic
to be put in package.json. Eventually this can be moved
to a test helper class instead.

MatthiasKainer committed:
Added states for consumer-c

MatthiasKainer committed:
Use Pact Data as test fixture
